File: binding/go/leopard_test.go
// Copyright 2022-2023 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is
// located in the "LICENSE" file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing permissions and
// limitations under the License.
//

package leopard

import (
	"encoding/binary"
	"encoding/json"
	"flag"
	"io/ioutil"
	"log"
	"math"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"

	"github.com/agnivade/levenshtein"
)

type LanguageTestParameters struct {
	language                  string
	testAudioFile             string
	transcript                string
	transcriptWithPunctuation string
	errorRate                 float32
	words                     []LeopardWord
}

type DiarizationTestParameters struct {
	language      string
	testAudioFile string
	words         []LeopardWord
}

var (
	testAccessKey    string
	leopard          Leopard
	languageTests    []LanguageTestParameters
	diarizationTests []DiarizationTestParameters
)

func TestMain(m *testing.M) {

	flag.StringVar(&testAccessKey, "access_key", "", "AccessKey for testing")
	flag.Parse()

	languageTests, diarizationTests = loadTestData()
	os.Exit(m.Run())
}

func isClose(value, expected, tolerance float32) bool {
	return math.Abs(float64(value-expected)) <= float64(tolerance)
}

func appendLanguage(s string, language string) string {
	if language == "en" {
		return s
	} else {
		return s + "_" + language
	}
}

func loadTestData() (languageTests []LanguageTestParameters, diarizationTests []DiarizationTestParameters) {

	content, err := ioutil.ReadFile("../../resources/.test/test_data.json")
	if err != nil {
		log.Fatalf("Could not read test data json: %v", err)
	}

	var testData struct {
		Tests struct {
			LanguageTests []struct {
				Language                  string  `json:"language"`
				AudioFile                 string  `json:"audio_file"`
				Transcript                string  `json:"transcript"`
				TranscriptWithPunctuation string  `json:"transcript_with_punctuation"`
				ErrorRate                 float32 `json:"error_rate"`
				Words                     []struct {
					Word       string  `json:"word"`
					StartSec   float32 `json:"start_sec"`
					EndSec     float32 `json:"end_sec"`
					Confidence float32 `json:"confidence"`
					SpeakerTag int32   `json:"speaker_tag"`
				} `json:"words"`
			} `json:"language_tests"`
			DiarizationTests []struct {
				Language  string `json:"language"`
				AudioFile string `json:"audio_file"`
				Words     []struct {
					Word       string `json:"word"`
					SpeakerTag int32  `json:"speaker_tag"`
				} `json:"words"`
			} `json:"diarization_tests"`
		} `json:"tests"`
	}
	err = json.Unmarshal(content, &testData)
	if err != nil {
		log.Fatalf("Could not decode test data json: %v", err)
	}

	for _, x := range testData.Tests.LanguageTests {
		languageTestParameters := LanguageTestParameters{
			language:                  x.Language,
			testAudioFile:             x.AudioFile,
			transcript:                x.Transcript,
			transcriptWithPunctuation: x.TranscriptWithPunctuation,
			errorRate:                 x.ErrorRate,
		}

		for _, y := range x.Words {
			word := LeopardWord{
				Word:       y.Word,
				StartSec:   y.StartSec,
				EndSec:     y.EndSec,
				Confidence: y.Confidence,
				SpeakerTag: y.SpeakerTag,
			}
			languageTestParameters.words = append(languageTestParameters.words, word)
		}

		languageTests = append(languageTests, languageTestParameters)
	}

	for _, x := range testData.Tests.DiarizationTests {
		diarizationTestParameters := DiarizationTestParameters{
			language:      x.Language,
			testAudioFile: x.AudioFile,
		}

		for _, y := range x.Words {
			word := LeopardWord{
				Word:       y.Word,
				SpeakerTag: y.SpeakerTag,
			}
			diarizationTestParameters.words = append(diarizationTestParameters.words, word)
		}

		diarizationTests = append(diarizationTests, diarizationTestParameters)
	}

	return languageTests, diarizationTests
}

func validateMetadata(t *testing.T, referenceWords []LeopardWord, words []LeopardWord, enableDiarization bool) {
	for i := range words {
		word := strings.ToUpper(words[i].Word)
		referenceWord := strings.ToUpper(referenceWords[i].Word)
		if word != referenceWord {
			t.Fatalf("Word `%s` did not match expected word `%s`", word, referenceWord)
		}
		if !isClose(words[i].StartSec, referenceWords[i].StartSec, 0.1) {
			t.Fatalf("Word %d started at %f, expected %f", i, words[i].StartSec, referenceWords[i].StartSec)
		}
		if !isClose(words[i].EndSec, referenceWords[i].EndSec, 0.1) {
			t.Fatalf("Word %d ended at %f, expected %f", i, words[i].EndSec, referenceWords[i].EndSec)
		}
		if !isClose(words[i].Confidence, referenceWords[i].Confidence, 0.1) {
			t.Fatalf("Word %d had a confidence of %f, expected %f", i, words[i].Confidence, referenceWords[i].Confidence)
		}
		if enableDiarization {
			if words[i].SpeakerTag != referenceWords[i].SpeakerTag {
				t.Fatalf("Word %d had speaker_tag of %d, expected %d", i, words[i].SpeakerTag, referenceWords[i].SpeakerTag)
			}
		} else {
			if words[i].SpeakerTag != -1 {
				t.Fatalf("Word %d had speaker_tag of %d, expected -1", i, words[i].SpeakerTag)
			}
		}
	}
}

func TestVersion(t *testing.T) {
	leopard = NewLeopard(testAccessKey)
	leopard.EnableAutomaticPunctuation = true
	err := leopard.Init()
	if err != nil {
		log.Fatalf("Failed to init leopard with: %v", err)
	}
	defer leopard.Delete()

	if reflect.TypeOf(Version).Name() != "string" {
		t.Fatal("Unexpected version format.")
	}
	if Version == "" {
		t.Fatal("Failed to get version.")
	}
}

func runProcessTestCase(
	t *testing.T,
	language string,
	testAudioFile string,
	referenceTranscript string,
	targetErrorRate float32,
	enableAutomaticPunctuation bool,
	enableDiarization bool,
	referenceWords []LeopardWord) {

	leopard = NewLeopard(testAccessKey)
	leopard.EnableAutomaticPunctuation = enableAutomaticPunctuation
	leopard.EnableDiarization = enableDiarization

	modelPath, _ := filepath.Abs(filepath.Join("../../lib/common", appendLanguage("leopard_params", language)+".pv"))
	leopard.ModelPath = modelPath

	err := leopard.Init()
	if err != nil {
		log.Fatalf("Failed to init leopard with: %v", err)
	}
	defer leopard.Delete()

	testAudioPath, _ := filepath.Abs(filepath.Join("../../resources/audio_samples", testAudioFile))
	data, err := ioutil.ReadFile(testAudioPath)
	if err != nil {
		t.Fatalf("Could not read test file: %v", err)
	}
	data = data[44:] // skip header

	totalFrames := len(data) / 2
	pcm := make([]int16, totalFrames)

	for i := 0; i < totalFrames; i++ {
		pcm[i] = int16(binary.LittleEndian.Uint16(data[2*i : (2*i)+2]))
	}

	transcript, words, err := leopard.Process(pcm)
	if err != nil {
		t.Fatalf("Failed to process pcm buffer: %v", err)
	}

	errorRate := float32(levenshtein.ComputeDistance(transcript, referenceTranscript)) / float32(len(referenceTranscript))
	if errorRate >= targetErrorRate {
		t.Fatalf("Expected '%f' got '%f'", targetErrorRate, errorRate)
	}

	validateMetadata(t, referenceWords, words, enableDiarization)
}

func runProcessFileTestCase(
	t *testing.T,
	language string,
	testAudioFile string,
	referenceTranscript string,
	targetErrorRate float32,
	enableAutomaticPunctuation bool,
	enableDiarization bool,
	referenceWords []LeopardWord) {

	leopard = NewLeopard(testAccessKey)
	leopard.EnableAutomaticPunctuation = enableAutomaticPunctuation
	leopard.EnableDiarization = enableDiarization

	modelPath, _ := filepath.Abs(filepath.Join("../../lib/common", appendLanguage("leopard_params", language)+".pv"))
	leopard.ModelPath = modelPath

	err := leopard.Init()
	if err != nil {
		log.Fatalf("Failed to init leopard with: %v", err)
	}
	defer leopard.Delete()

	testAudioPath, _ := filepath.Abs(filepath.Join("../../resources/audio_samples", testAudioFile))
	transcript, words, err := leopard.ProcessFile(testAudioPath)
	if err != nil {
		t.Fatalf("Failed to process pcm buffer: %v", err)
	}
	errorRate := float32(levenshtein.ComputeDistance(transcript, referenceTranscript)) / float32(len(referenceTranscript))
	if errorRate >= targetErrorRate {
		t.Fatalf("Expected '%f' got '%f'", targetErrorRate, errorRate)
	}

	validateMetadata(t, referenceWords, words, enableDiarization)
}

func runDiarizationTestCase(
	t *testing.T,
	language string,
	testAudioFile string,
	referenceWords []LeopardWord) {

	leopard = NewLeopard(testAccessKey)
	leopard.EnableDiarization = true

	modelPath, _ := filepath.Abs(filepath.Join("../../lib/common", appendLanguage("leopard_params", language)+".pv"))
	leopard.ModelPath = modelPath

	err := leopard.Init()
	if err != nil {
		log.Fatalf("Failed to init leopard with: %v", err)
	}
	defer leopard.Delete()

	testAudioPath, _ := filepath.Abs(filepath.Join("../../resources/audio_samples", testAudioFile))
	_, words, err := leopard.ProcessFile(testAudioPath)
	if err != nil {
		t.Fatalf("Failed to process pcm buffer: %v", err)
	}

	for i := range words {
		word := strings.ToUpper(words[i].Word)
		referenceWord := strings.ToUpper(referenceWords[i].Word)
		if word != referenceWord {
			t.Fatalf("Word `%s` did not match expected word `%s`", word, referenceWord)
		}
		if words[i].SpeakerTag != referenceWords[i].SpeakerTag {
			t.Fatalf("Word %d had speaker_tag of %d, expected %d", i, words[i].SpeakerTag, referenceWords[i].SpeakerTag)
		}
	}
}

func TestProcess(t *testing.T) {
	for _, test := range languageTests {
		t.Logf("Running process data test for `%s`", test.language)
		runProcessTestCase(
			t,
			test.language,
			test.testAudioFile,
			test.transcript,
			test.errorRate,
			false,
			false,
			test.words)
	}
}

func TestProcessFile(t *testing.T) {
	for _, test := range languageTests {
		t.Logf("Running process file test for `%s`", test.language)
		runProcessFileTestCase(
			t,
			test.language,
			test.testAudioFile,
			test.transcript,
			test.errorRate,
			false,
			false,
			test.words)
	}
}

func TestProcessFileWithPunctuation(t *testing.T) {
	for _, test := range languageTests {
		t.Logf("Running process file with punctuation test for `%s`", test.language)
		runProcessFileTestCase(
			t,
			test.language,
			test.testAudioFile,
			test.transcriptWithPunctuation,
			test.errorRate,
			true,
			false,
			test.words)
	}
}

func TestProcessFileWithDiarization(t *testing.T) {
	for _, test := range languageTests {
		t.Logf("Running process file with diarization test for `%s`", test.language)
		runProcessFileTestCase(
			t,
			test.language,
			test.testAudioFile,
			test.transcript,
			test.errorRate,
			false,
			true,
			test.words)
	}
}

func TestDiarization(t *testing.T) {
	for _, test := range diarizationTests {
		t.Logf("Running diarization test for `%s`", test.language)
		runDiarizationTestCase(
			t,
			test.language,
			test.testAudioFile,
			test.words)
	}
}

func TestProcessEmptyFile(t *testing.T) {
	leopard = NewLeopard(testAccessKey)
	err := leopard.Init()
	if err != nil {
		log.Fatalf("Failed to init leopard with: %v", err)
	}
	defer leopard.Delete()

	testAudioPath, _ := filepath.Abs(filepath.Join("../../resources/audio_samples", "empty.wav"))
	transcript, words, err := leopard.ProcessFile(testAudioPath)
	if err != nil {
		t.Fatalf("Failed to process pcm buffer: %v", err)
	}
	if len(transcript) > 0 {
		t.Fatalf("Leopard returned transcript on empty file `%s`", transcript)
	}

	if len(words) > 0 {
		t.Fatalf("Leopard returned %d words on empty file", len(words))
	}
}

File: binding/java/test/ai/picovoice/leopard/LeopardTest.java
/*
    Copyright 2022-2023 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.leopard;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import org.apache.commons.text.similarity.LevenshteinDistance;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.stream.Stream;

import java.util.Locale;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class LeopardTest {
    private final String accessKey = System.getProperty("pvTestingAccessKey");
    private Leopard leopard;

    private static String appendLanguage(String s, String language) {
        if (language.equals("en")) {
            return s;
        }
        return s + "_" + language;
    }

    private static float getErrorRate(String transcript, String expectedTranscript) {
        return (float) LevenshteinDistance
                .getDefaultInstance()
                .apply(transcript, expectedTranscript) / (float) expectedTranscript.length();
    }

    private static ProcessTestData[] loadProcessTestData() throws IOException {
        final Path testDataPath = Paths.get(System.getProperty("user.dir"))
                .resolve("../../resources/.test")
                .resolve("test_data.json");
        final String testDataContent = new String(Files.readAllBytes(testDataPath), StandardCharsets.UTF_8);
        final JsonObject testDataJson = JsonParser.parseString(testDataContent).getAsJsonObject();

        final JsonArray testParameters = testDataJson
                .getAsJsonObject("tests")
                .getAsJsonArray("language_tests");

        final ProcessTestData[] processTestData = new ProcessTestData[testParameters.size()];
        for (int i = 0; i < testParameters.size(); i++) {
            final JsonObject testData = testParameters.get(i).getAsJsonObject();
            final String language = testData.get("language").getAsString();
            final String testAudioFile = testData.get("audio_file").getAsString();
            final String transcript = testData.get("transcript").getAsString();
            final String transcriptWithPunctuation = testData.get("transcript_with_punctuation").getAsString();
            final float errorRate = testData.get("error_rate").getAsFloat();

            final JsonArray wordsJson = testData.getAsJsonArray("words");
            final LeopardTranscript.Word[] words = new LeopardTranscript.Word[wordsJson.size()];
            for (int j = 0; j < wordsJson.size(); j++) {
                final JsonObject wordJson = wordsJson.get(j).getAsJsonObject();
                words[j] = new LeopardTranscript.Word(
                        wordJson.get("word").getAsString(),
                        wordJson.get("confidence").getAsFloat(),
                        wordJson.get("start_sec").getAsFloat(),
                        wordJson.get("end_sec").getAsFloat(),
                        wordJson.get("speaker_tag").getAsInt());
            }
            processTestData[i] = new ProcessTestData(
                    language,
                    testAudioFile,
                    transcript,
                    transcriptWithPunctuation,
                    errorRate,
                    words);
        }
        return processTestData;
    }

    private static DiarizationTestData[] loadDiarizationTestData() throws IOException {
        final Path testDataPath = Paths.get(System.getProperty("user.dir"))
                .resolve("../../resources/.test")
                .resolve("test_data.json");
        final String testDataContent = new String(Files.readAllBytes(testDataPath), StandardCharsets.UTF_8);
        final JsonObject testDataJson = JsonParser.parseString(testDataContent).getAsJsonObject();

        final JsonArray testParameters = testDataJson
                .getAsJsonObject("tests")
                .getAsJsonArray("diarization_tests");

        final DiarizationTestData[] diarizationTestData = new DiarizationTestData[testParameters.size()];
        for (int i = 0; i < testParameters.size(); i++) {
            final JsonObject testData = testParameters.get(i).getAsJsonObject();
            final String language = testData.get("language").getAsString();
            final String testAudioFile = testData.get("audio_file").getAsString();

            final JsonArray wordsJson = testData.getAsJsonArray("words");
            final LeopardTranscript.Word[] words = new LeopardTranscript.Word[wordsJson.size()];
            for (int j = 0; j < wordsJson.size(); j++) {
                final JsonObject wordJson = wordsJson.get(j).getAsJsonObject();
                words[j] = new LeopardTranscript.Word(
                        wordJson.get("word").getAsString(),
                        0,
                        0,
                        0,
                        wordJson.get("speaker_tag").getAsInt());
            }
            diarizationTestData[i] = new DiarizationTestData(
                    language,
                    testAudioFile,
                    words);
        }
        return diarizationTestData;
    }

    private static short[] readAudioFile(String testAudioPath) throws Exception {

        File testAudioFile = new File(testAudioPath);
        AudioInputStream audioInputStream = AudioSystem.getAudioInputStream(testAudioFile);

        int byteDepth = audioInputStream.getFormat().getFrameSize();
        int frameLength = (int) audioInputStream.getFrameLength();
        byte[] bytes = new byte[frameLength * byteDepth];
        short[] pcm = new short[frameLength];

        audioInputStream.read(bytes);
        ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer().get(pcm);
        return pcm;
    }

    private static Stream<Arguments> processTestProvider() throws IOException {
        final ProcessTestData[] processTestData = loadProcessTestData();
        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (ProcessTestData processTestDataItem : processTestData) {
            testArgs.add(Arguments.of(
                    processTestDataItem.language,
                    processTestDataItem.audioFile,
                    processTestDataItem.transcript,
                    false,
                    false,
                    processTestDataItem.errorRate,
                    processTestDataItem.words));
        }

        return testArgs.stream();
    }

    private static Stream<Arguments> processFileTestProvider() throws IOException {
        final ProcessTestData[] processTestData = loadProcessTestData();
        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (ProcessTestData processTestDataItem : processTestData) {
            testArgs.add(Arguments.of(
                    processTestDataItem.language,
                    processTestDataItem.audioFile,
                    processTestDataItem.transcript,
                    false,
                    false,
                    processTestDataItem.errorRate,
                    processTestDataItem.words));
            testArgs.add(Arguments.of(
                    processTestDataItem.language,
                    processTestDataItem.audioFile,
                    processTestDataItem.transcriptWithPunctuation,
                    true,
                    false,
                    processTestDataItem.errorRate,
                    processTestDataItem.words));
            testArgs.add(Arguments.of(
                    processTestDataItem.language,
                    processTestDataItem.audioFile,
                    processTestDataItem.transcript,
                    false,
                    true,
                    processTestDataItem.errorRate,
                    processTestDataItem.words));
        }

        return testArgs.stream();
    }

    private static Stream<Arguments> diarizationTestProvider() throws IOException {
        final DiarizationTestData[] diarizationTestData = loadDiarizationTestData();
        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (DiarizationTestData diarizationTestDataItem : diarizationTestData) {
            testArgs.add(Arguments.of(
                    diarizationTestDataItem.language,
                    diarizationTestDataItem.audioFile,
                    diarizationTestDataItem.words));
        }

        return testArgs.stream();
    }

    @AfterEach
    void tearDown() {
        leopard.delete();
    }

    void validateMetadata(
            LeopardTranscript.Word[] words,
            LeopardTranscript.Word[] referenceWords,
            boolean enableDiarization) {
        assertEquals(words.length, referenceWords.length);
        for (int i = 0; i < words.length; i++) {
            assertEquals(
                    words[i].getWord().toUpperCase(Locale.ENGLISH),
                    referenceWords[i].getWord().toUpperCase(Locale.ENGLISH)
            );
            assertEquals(words[i].getStartSec(), referenceWords[i].getStartSec(), 0.1);
            assertEquals(words[i].getEndSec(), referenceWords[i].getEndSec(), 0.1);
            assertEquals(words[i].getConfidence(), referenceWords[i].getConfidence(), 0.1);
            if (enableDiarization) {
                assertEquals(words[i].getSpeakerTag(), referenceWords[i].getSpeakerTag());
            } else {
                assertEquals(words[i].getSpeakerTag(), -1);
            }
        }
    }

    @Test
    void getVersion() throws Exception {
        leopard = new Leopard.Builder()
                .setAccessKey(accessKey)
                .build();

        assertTrue(leopard.getVersion() != null && !leopard.getVersion().equals(""));
    }

    @Test
    void getSampleRate() throws Exception {
        leopard = new Leopard.Builder()
                .setAccessKey(accessKey)
                .build();
        assertTrue(leopard.getSampleRate() > 0);
    }

    @ParameterizedTest(name = "test process data for ''{0}'' with punctuation ''{3}'' and diarization ''{4}''")
    @MethodSource("processTestProvider")
    void process(
            String language,
            String testAudioFile,
            String referenceTranscript,
            boolean enableAutomaticPunctuation,
            boolean enableDiarization,
            float targetErrorRate,
            LeopardTranscript.Word[] referenceWords) throws Exception {
        String modelPath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../lib/common/%s.pv", appendLanguage("leopard_params", language)))
                .toString();

        leopard = new Leopard.Builder()
                .setAccessKey(accessKey)
                .setModelPath(modelPath)
                .setEnableAutomaticPunctuation(enableAutomaticPunctuation)
                .setEnableDiarization(enableDiarization)
                .build();

        String testAudioPath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../resources/audio_samples/%s", testAudioFile))
                .toString();
        short[] pcm = readAudioFile(testAudioPath);
        LeopardTranscript result = leopard.process(pcm);

        assertTrue(getErrorRate(result.getTranscriptString(), referenceTranscript) < targetErrorRate);
        validateMetadata(
                result.getWordArray(),
                referenceWords,
                enableDiarization);
    }

    @ParameterizedTest(name = "test process file for ''{0}'' with punctuation ''{3}'' and diarization ''{4}''")
    @MethodSource("processFileTestProvider")
    void processFile(
            String language,
            String testAudioFile,
            String referenceTranscript,
            boolean enableAutomaticPunctuation,
            boolean enableDiarization,
            float targetErrorRate,
            LeopardTranscript.Word[] referenceWords) throws Exception {

        String modelPath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../lib/common/%s.pv", appendLanguage("leopard_params", language)))
                .toString();

        leopard = new Leopard.Builder()
                .setAccessKey(accessKey)
                .setModelPath(modelPath)
                .setEnableAutomaticPunctuation(enableAutomaticPunctuation)
                .setEnableDiarization(enableDiarization)
                .build();

        String testAudioPath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../resources/audio_samples/%s", testAudioFile))
                .toString();

        LeopardTranscript result = leopard.processFile(testAudioPath);
        assertTrue(getErrorRate(result.getTranscriptString(), referenceTranscript) < targetErrorRate);
        validateMetadata(
                result.getWordArray(),
                referenceWords,
                enableDiarization);
    }

    @ParameterizedTest(name = "test diarization for ''{0}''")
    @MethodSource("diarizationTestProvider")
    void diarization(
            String language,
            String testAudioFile,
            LeopardTranscript.Word[] referenceWords) throws Exception {

        String modelPath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../lib/common/%s.pv", appendLanguage("leopard_params", language)))
                .toString();

        leopard = new Leopard.Builder()
                .setAccessKey(accessKey)
                .setModelPath(modelPath)
                .setEnableDiarization(true)
                .build();

        String testAudioPath = Paths.get(System.getProperty("user.dir"))
                .resolve(String.format("../../resources/audio_samples/%s", testAudioFile))
                .toString();

        LeopardTranscript result = leopard.processFile(testAudioPath);
        LeopardTranscript.Word[] words = result.getWordArray();
        assertEquals(result.getWordArray().length, referenceWords.length);
        for (int i = 0; i < words.length; i++) {
            assertEquals(
                    words[i].getWord().toUpperCase(Locale.ENGLISH),
                    referenceWords[i].getWord().toUpperCase(Locale.ENGLISH)
            );
            assertEquals(words[i].getSpeakerTag(), referenceWords[i].getSpeakerTag());
        }
    }

    private static class ProcessTestData {
        public final String language;
        public final String audioFile;
        public final String transcript;
        public final String transcriptWithPunctuation;
        public final float errorRate;
        public final LeopardTranscript.Word[] words;

        public ProcessTestData(
                String language,
                String audioFile,
                String transcript,
                String transcriptWithPunctuation,
                float errorRate,
                LeopardTranscript.Word[] words) {
            this.language = language;
            this.audioFile = audioFile;
            this.transcript = transcript;
            this.transcriptWithPunctuation = transcriptWithPunctuation;
            this.errorRate = errorRate;
            this.words = words;
        }
    }

    private static class DiarizationTestData {
        public final String language;
        public final String audioFile;
        public final LeopardTranscript.Word[] words;

        public DiarizationTestData(
                String language,
                String audioFile,
                LeopardTranscript.Word[] words) {
            this.language = language;
            this.audioFile = audioFile;
            this.words = words;
        }
    }
}
